### PR TITLE
Added new transforms submodules and fill_scale_tril.py transform to c…

### PR DESCRIPTION
**Pull request by LuisA92**
This PR adds a `transforms` submodule directory as well as a 'fill_scale_tril.py' transform to convert vectors into a lower triangular matrix with a positive diagonal. The current implementation uses a `SoftplusTransform' to constraint the diagonal. This should be compatible with [torch.distributions.transforms.ComposeTransform](https://pytorch.org/docs/stable/distributions.html#torch.distributions.transforms.ComposeTransform), but I have not tested it.

### CODE REVIEW

**Comment by LuisA92** [comment]
The code is failing because [torch.distributions.transforms.ComposeTransform](https://pytorch.org/docs/stable/distributions.html#torch.distributions.transforms.ComposeTransform) requires me to define the variables `domain`, `codomain`, and `bijective` but they are not used in the class.

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import torch
+from torch.distributions import Transform, constraints
+from torch.distributions.transforms import SoftplusTransform
+from torch.distributions.utils import vec_to_tril_matrix, tril_matrix_to_vec
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py:4]: I have been looking all over for this method. How did you even find this?

Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py:4]: I have been looking all over for this method. How did you even find this?

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        Args:
+            diag_transform: transformation used on diagonal to ensure positive values.
+            Default is SoftplusTransform
+            diag_shift (float): small offset to avoid diagonals very close to zero.
+            Default offset is 1e-05
+
+        """
+        super().__init__()
+        self.diag_transform = (
+            diag_transform if diag_transform is not None else SoftplusTransform()
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py]: this is not a bug, but we should discuss as a group whether we want to use ExpTransform or SofplusTransform in cases like these. either way, you might consider refactoring this to grab the appropriate constraint for positivity from the constraints registry. (https://pytorch.org/docs/stable/distributions.html#module-torch.distributions.constraint_registry)

Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: Following up on this, the diag shift stuff __should__ be included in the constraint registry for positive constraints. however, i think the default bijection for positive variables doesn't include an offset.

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            x (torch.Tensor): Input vector to transform
+        Returns:
+            torch.Tensor: Transformed lower triangular matrix
+        """
+        x = vec_to_tril_matrix(x)
+        diagonal_elements = x.diagonal(dim1=-2, dim2=-1)
+        transformed_diagonal = self.diag_transform(diagonal_elements)
+        if self.diag_shift is not None:
+            transformed_diagonal += self.diag_shift
+        x.diagonal(dim1=-2, dim2=-1).copy_(transformed_diagonal)
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py]: i don't understand this line. does it do anything?

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        else:
+            transformed_diagonal = self.diag_transform.inv(diagonal_elements)
+        y.diagonal(dim1=-2, dim2=-1).copy_(transformed_diagonal)
+        return tril_matrix_to_vec(y)
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Computes the log absolute determinant of the Jacobian matrix for the transformation.
+
+        Assumes that Softplus is used on the diagonal.
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py]: this is too restrictive. we should allow any transform to be used here.

Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: you should be able to get the derivatives of self.diag_transform programmatically.

File: tests/transforms/fill_scale_tril.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import pytest
+from rs_distributions.transforms.fill_scale_tril import FillScaleTriL
+import torch
+from torch.distributions.constraints import lower_cholesky
+
+
+@pytest.mark.parametrize("input_shape", [(6,), (10,)])
```
Comment (kmdalton) [inline tests/transforms/fill_scale_tril.py]: this test should include nested batches like
```
(a, b, c, d*(d-1)/2) -> (a, b, c, d, d)
```

File: tests/transforms/fill_scale_tril.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    assert isinstance(transformed_vector, torch.Tensor)
+    assert transformed_vector.shape == (
+        (-1 + torch.sqrt(torch.tensor(1 + input_shape[0] * 8))) / 2,
+        (-1 + torch.sqrt(torch.tensor(1 + input_shape[0] * 8))) / 2,
+    )
+    assert lower_cholesky.check(transformed_vector)
+
+
+@pytest.mark.parametrize("input_vector", [torch.randn(3), torch.randn(6)])
```
Comment (kmdalton) [inline tests/transforms/fill_scale_tril.py]: nested batches

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        """
+        diagonal_elements = y.diagonal(dim1=-2, dim2=-1)
+        if self.diag_shift is not None:
+            transformed_diagonal = self.diag_transform.inv(
+                diagonal_elements - self.diag_shift
+            )
+        else:
+            transformed_diagonal = self.diag_transform.inv(diagonal_elements)
+        y.diagonal(dim1=-2, dim2=-1).copy_(transformed_diagonal)
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py]: this line doesn't make sense to me. is it necessary?

**Review by kmdalton — CHANGES_REQUESTED** [review]
this looks like a good start. i think the biggest things that need addressing are
1) support ComposeTransform and add a test to make sure it works (maybe this is not possible for some reason?). 
2) support bijectors other than softplus for the diagonal. i maintain that hardcoding the derivatives in the log_det_jacobian block is too restrictive. it should be relatively easy to get the necessary info from self.diag_transform

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 1 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py:4]: I saw that it was used in the source code for the [CorrCholeskyTransform](https://pytorch.org/docs/stable/distributions.html#torch.distributions.transforms.CorrCholeskyTransform)

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 3 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: this line copies the `transformed_diagonal` back onto the diagonal of `x`

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 4 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: Okay I will modify to get derivatives of `self.diag_transform`

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 7 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: This copies the `transformed_diagonal` back onto `y`

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 3 above)
Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: i read the docstring and i'm still confused. specifically, i don't know how seriously to take the word "copy" in this context. is this generating new objects that use extra memory? do they retain their relationship to the rest of the computational graph? i don't think we want  to copy the diagonal, right? we just want to assign its transformed values back into the array. doesn't it make more sense to do something like this?
```python
diag = x.diagonal(dim1=-2, dim2=-1)
result = x.diagonal_scatter(
    constraint_function(diag), dim1=-2, dim2=-1
)
```

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 4 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: Im not entirely sure if this is the correct way to do it, but this is what I have so far: 

```python 

    def log_abs_det_jacobian(self, x, y):
        L = vec_to_tril_matrix(x)
        diag = L.diagonal()
        diag.requires_grad_(True)
        transformed_diag = self.diag_transform(diag + self.diag_shift)
        derivatives = torch.autograd.grad(
            outputs=transformed_diag,
            inputs=diag,
            grad_outputs=torch.ones_like(transformed_diag),
        )[0]
        log_det_jacobian = torch.log(torch.abs(derivatives)).sum()
        return log_det_jacobian
```
When I use `SoftplusTransform` for `self.diag_transform` this method matches the gradients if I use: 

```python 
transform = FillScaleTriL()
softplus_transform = SoftplusTransform()

x = torch.randn(3, requires_grad=True)
L = transform._call(x)

# using analytical form
L_ = vec_to_tril_matrix(x)
diag = L_.diagonal()
derivatives = torch.sigmoid(diag)
result1 = torch.log(torch.abs(derivatives)).sum()

# using FillScaleTriL.log_abs_det_jacobian
result2 = transform.log_abs_det_jacobian(x,L)

#checking if gradients match
torch.allclose(result1,result2)
```

File: tests/transforms/fill_scale_tril.py (hunk 5 above)
Comment (LuisA92) [reply tests/transforms/fill_scale_tril.py]: I modified the tests to handle nested batches.

File: tests/transforms/fill_scale_tril.py (hunk 6 above)
Comment (LuisA92) [reply tests/transforms/fill_scale_tril.py]: I modified the tests to handle nested batches.

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 3 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: You're right. Using .copy_() did _not_ retain the computational graph but `.diagonal_scatter()` does.

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 2 above)
Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: Why don't you use the constraint registry to populate the diagonal transform if it is not specified? You can do this with `torch.distributions.tranform_to(constraint)`.

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 4 above)
Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: I have some thoughts about this. In tfp, the `FillScaleTriL` bijector is [implemented](https://github.com/tensorflow/probability/blob/65f265c62bb1e2d15ef3e25104afb245a6d52429/tensorflow_probability/python/bijectors/fill_scale_tril.py#L132) as a `Chain` which is their equivalent of the `ComposeTransform` class in `torch`.  Specifically, it is the composition of two bijectors, `FillTriL` and `TransformDiagonal`. Here are the Jacobian implementations for those base classes:
 - [FillTriL](https://github.com/tensorflow/probability/blob/65f265c62bb1e2d15ef3e25104afb245a6d52429/tensorflow_probability/python/bijectors/fill_triangular.py#L97)
 - [TransformDiagonal](https://github.com/tensorflow/probability/blob/65f265c62bb1e2d15ef3e25104afb245a6d52429/tensorflow_probability/python/bijectors/transform_diagonal.py#L81)
 
The `FillTriL` is pretty trivial because it essentially amounts to reshaping the input in a funny way. The TransformDiagonal Jacobian is __very__ straightforward to implement but less obvious. The developer left a nice note explaining it. 
 
 My suggestion is that you adopt this structure for your own implementation. Write your own `FillTriL` and `TransformDiagonal` transforms and implement `FillScaleTriL` as a `ComposeTransform`

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 2 above)
Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: i think having a separate "diag_shift" argument is sloppy. why not make the default diag_transform
```python
self.diag_transform = torch.distributions.ComposeTransform((
    torch.distributions.SoftplusTransform(),
    torch.distributions.AffineTransform(1e-5, 1.),
))
```
? if people want a different shift, they can supply it by the transform. thoughts?

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
+import torch
+from torch.distributions import Transform, constraints
+from torch.distributions.transforms import SoftplusTransform
+from torch.distributions.utils import vec_to_tril_matrix, tril_matrix_to_vec
+
+
+class FillScaleTriL(Transform):
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py]: I request you add a `params_size` staticmethod a la tfp's multivariate normal: https://www.tensorflow.org/probability/api_docs/python/tfp/layers/MultivariateNormalTriL#params_size

**Review by kmdalton — CHANGES_REQUESTED** [review]
I'm requesting a couple of changes. First, can we have a `params_size` staticmethod attached to the FillScaleTriL transform? that why no one has to remember the formula for the vector size.  Second, I don't like mixing hardcoded `diag_shift` parameters with transforms. if you want your bijector to have a shift, it should just be in the bijector.

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 2 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: This is an older commit, i have removed the `diag_shift` in the updated version

Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: My bad, I see what you did now. I think it is good to keep the shift in the default but implement it as part of the transform. I would recommend the following pattern:
```python
    def __init__(self, diag_transform=None):
        if diag_transform is None:
            diag_transform = torch.distributions.ComposeTransform((
                 torch.distributions.SoftplusTransform(),
                 torch.distributions.AffineTransform(1e-5, 1.),
             ))
        super().__init__([FillTriL(), DiagTransform(diag_transform=diag_transform)])
        self.diag_transform = diag_transform
```

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 8 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: Something like 

```python
@staticmethod
def params_size(event_size):
# for lower triangular we need n*(n+1)//2 elements
return event_size * (event_size + 1) // 2 
```

Comment (kmdalton) [reply src/rs_distributions/transforms/fill_scale_tril.py]: yeah that's what i am thinking. it's helpful to have when you are in dimensions greater than 3.

File: src/rs_distributions/transforms/fill_scale_tril.py (hunk 2 above)
Comment (LuisA92) [reply src/rs_distributions/transforms/fill_scale_tril.py]: Okay I added this to `FillScaleTriL`

File: src/rs_distributions/transforms/fill_scale_tril.py [hunk 9]
```diff
@@ -106,3 +111,17 @@ def log_abs_det_jacobian(self, x, y):
         x = FillTriL()._call(x)
         diagonal = x.diagonal(dim1=-2, dim2=-1)
         return self.diag_transform.log_abs_det_jacobian(diagonal, diagonal)
+
+    @staticmethod
+    def params_size(event_size):
+        """
+        Returns the number of parameters required to create lower triangular matrix, which is given by n*(n+1)//2
```
Comment (kmdalton) [inline src/rs_distributions/transforms/fill_scale_tril.py]: Docstring suggestion: ```Returns the number of parameters required to create an n-by-n lower triangular matrix...```

**Review by kmdalton — APPROVED** [review]
Looks good! Thanks
